### PR TITLE
Null safety

### PR DESCRIPTION
**Pull request by tim-smart**
Things to note:

* Doesn't try to be backwards compatible

### CODE REVIEW

File: functional_widget/lib/function_to_widget_class.dart [hunk 1]
```diff
@@ -250,23 +251,27 @@ class FunctionalWidgetGenerator
...
     return Constructor(
       (b) => b
         ..constant = true
         ..optionalParameters.add(Parameter((b) => b
+          ..required = keyIsRequired
           ..named = true
           ..name = 'key'
           ..docs.clear()
-          ..type = _keyRef))
+          ..type = keyIsRequired ? _keyRef : _optionalKeyRef))
```
Comment (tim-smart) [inline functional_widget/lib/function_to_widget_class.dart:274]: Better way of doing this?

File: functional_widget/lib/src/parameters.dart [hunk 2]
```diff
@@ -63,28 +61,25 @@ Reference _parameterToReference(ParameterElement element) {
...
 }
 
 FunctionType _functionTypedElementToFunctionType(
   element_type.FunctionType element,
 ) {
   return FunctionType((b) {
-    return b
+    b
       ..returnType = _typeToReference(element.returnType)
       // TODO set appropriate `nullabilitySuffix`
```
Comment (tim-smart) [inline functional_widget/lib/src/parameters.dart:79]: Not sure what this is used for, so left it as is.

File: functional_widget/lib/function_to_widget_class.dart [hunk 3]
```diff
@@ -58,11 +59,11 @@ class FunctionalWidgetGenerator
         element: element,
       );
     }
-    var function = element as FunctionElement;
+    var function = element;
     if (function.isAsynchronous ||
         function.isExternal ||
         function.isGenerator ||
-        function.returnType?.getDisplayString(withNullability: false) !=
+        function.returnType.getDisplayString(withNullability: false) !=
```
Comment (tim-smart) [inline functional_widget/lib/function_to_widget_class.dart:66]: Need to set this to true

File: functional_widget/lib/function_to_widget_class.dart (hunk 1 above)
Comment (tim-smart) [reply functional_widget/lib/function_to_widget_class.dart:274]: I think a `TypeReference` builder would be the right thing to use here.

Can set `isNullable` to `keyIsRequired`.

File: functional_widget/lib/src/parameters.dart (hunk 2 above)
Comment (tim-smart) [reply functional_widget/lib/src/parameters.dart]: Turns out nullabilitySuffix does not need to be set here, at least I think so - after adding some more tests.

**Comment by rrousselGit** [comment]
Thanks!

I'll perform a few more tests and deploy it

**Pull request merged**